d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -833,4 +894,64 @@ private ImmutablePair<String, String> getZnodePathAndData(String cluster) {
   public boolean isWarmingUp() {
     return _isWarmingUp;
   }
+
+  private Map<Integer, PartitionData> validatePartitionData(Map<Integer, PartitionData> partitionData) {
+    if (_maxWeight == null) {
+      return partitionData;
+    }
+
+    Map<Integer, PartitionData> res = new HashMap<>(partitionData); // modifiable copy in case the input is unmodifiable
+    for (Map.Entry<Integer, PartitionData> entry : res.entrySet()) {
+      BigDecimal weight = BigDecimal.valueOf(entry.getValue().getWeight());
+      // check max weight
+      if (weight.compareTo(_maxWeight) > 0) {
+        _maxWeightBreachedCount.incrementAndGet();
+        switch (_actOnWeightBreach) {
+          case WARN:
+            _log.warn("", getMaxWeightBreachException(weight, entry.getKey()));
+            break;
+          case THROW:
+            throw getMaxWeightBreachException(weight, entry.getKey());
+          case RECTIFY:
+            entry.setValue(new PartitionData(_maxWeight.intValue()));
+            weight = _maxWeight;
+            _log.warn("Capped weight {} in Partition {} to the max weight allowed: {}.", weight.doubleValue(),
+                entry.getKey(), _maxWeight.intValue());
+            break;
+          case IGNORE:
+          default:
+            break;
+        }
+      }
+
+      // check decimal places
+      if (weight.scale() > _maxWeight.scale()) {
+        _weightDecimalPlacesBreachedCount.incrementAndGet();
+        switch (_actOnWeightBreach) {
+          case WARN: // both WARN and THROW only log the warning. Don't throw exception for decimal places.
+          case THROW:
+            _log.warn("", new IllegalArgumentException(String.format("Weight %f in Partition %d has more than %d"
+                + " decimal places. It will be rounded in the future.", weight.doubleValue(), entry.getKey(),
+                _maxWeight.scale())));
+            break;
+          case RECTIFY:
+            double newWeight = weight.setScale(_maxWeight.scale(), RoundingMode.HALF_UP).doubleValue();
+            entry.setValue(new PartitionData(newWeight));
+            _log.warn("Rounded weight {} in Partition {} to {} decimal places: {}.", weight.doubleValue(),
+                entry.getKey(), _maxWeight.scale(), newWeight);
+            break;
+          case IGNORE:
+          default:
+            break;
+        }
+      }
+    }
+    return res;
+  }
+
+  private IllegalArgumentException getMaxWeightBreachException(BigDecimal weight, int partition) {
+    return new IllegalArgumentException(String.format("[ACTION NEEDED] Weight %f in Partition %d is greater"
+        + " than the max weight allowed: %d. Please correct the weight. It will be force-capped to the max weight "
+        + "in the future.", weight.doubleValue(), partition, _maxWeight.intValue()));
```

Comment:
Technically you don't know if `BigDecimal` is an integer (even though in practice it'll only ever be 100). I think you can just have `than the max weight allowed: %s` and let the `toString` do the formatting right?

Reply:
yeah, good catch. fixed.